### DISCUSSION TITLE
Standard Output of "brew tap" is Empty When Run as a Sub-Process from a Separate Program

### DISCUSSION THREAD

**Discussion by buresdv** [Homebrew/discussions#4197] [Everyday usage]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 3.6.21
ORIGIN: https://github.com/Homebrew/brew
HEAD: 932d2cf3b77c9439a57b6a43577fc8d3b6399a62
Last commit: 4 days ago
Core tap origin: https://github.com/Homebrew/homebrew-core
Core tap HEAD: ca8f06d971c8332f146a1443e68d8a714daa0a51
Core tap last commit: 20 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.37.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.86.0 => /usr/bin/curl
macOS: 13.2-arm64
CLT: 14.2.0.0.1.1668646533
Xcode: 14.2 => /Applications/Xcode-14.2.0.app/Contents/Developer
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Your system is ready to brew.
```

### Description of issue

Hey everyone,

I don't know if this is a programming error on my part or something to do with Brew itself, but I have exhausted all other options, so I'm asking here.

I'm developing a helper GUI program for Brew for macOS. This program executes brew commands through sub-processes and then captures their outputs.

This works great with most commands like *brew install* and *brew update*. The process finishes and returns the output of the command as it should.

However, I have run into a problem with the commands *brew tap* and *brew untap*. For some reason, these two commands return absolutely nothing. When I execute them, I can see the command's real-time output in the terminal, but when I try to capture the final result through standard output, it's completely empty. To see what I mean, check out [this video](https://streamable.com/mnubl8).

As you can see, the command runs and outputs its usual output into the console. At the end, I select the line which starts with "Result", which should contain the full output of the command as it does with other commands, but as you can see, it's completely empty.

If you want to inspect the relevant code of my program, the function for executing and capturing command outputs is [here](https://github.com/buresdv/Cork/blob/main/Cork/Logic/Shell/Shell%20Interface.swift) and an example of running a command and capturing its output is [here](https://github.com/buresdv/Cork/blob/6342f5bf848f56a11144ccad562bc536144dd915/Cork/Logic/Upgrade%20Packages.swift#L21).

I have scoured Stack Overflow, Apple development forums and asked in various programming communities, but nobody knows why this is happening.

**Comment by gromgit** [comment]
> When I execute them, I can see the command's real-time output in the terminal, but when I try to capture the final result through standard output, it's completely empty.

Try capturing _standard error_ as well.

**Reply by buresdv** [reply]
We're making progress. Capturing standard error finally returned something, but it only seems to be the first output of the command. When I captured the standard error of the command *brew tap elastic/tap*, the pipe captured only "==> Tapping elastic/tap". For some reason, all other output of that command did not get captured.

So all of this output got ignored:
```
Cloning into '/opt/homebrew/Library/Taps/elastic/homebrew-tap'...
Tapped 17 formulae (37 files, 445.5KB).
```

**Reply by gromgit** [reply]
I don't do Swift, but it sounds like the problem is in your pipe handling. Try running the equivalent of `brew tap elastic/tap >& /tmp/tap.out` in your program, then read the `/tmp/tap.out` file.

**Reply by buresdv** [reply]
It seems like you were right. Those files include more of the output:
[tap.txt](https://github.com/Homebrew/discussions/files/10716074/tap.txt)
[untap.txt](https://github.com/Homebrew/discussions/files/10716076/untap.txt)

**Reply by gromgit** [reply]
Yeah, that's what I thought. I don't know how you're writing your code, and I don't Swift, but you should probably run the equivalent of `brew tap elastic/tap 2>&1` to redirect `stderr` to `stdout`, then you only have one pipe to worry about.
